teps/draft-20200911-hermekton.md

```diff
@@ -0,0 +1,214 @@
+That means that these are not a full security boundary.
+Hermetic builds should not be used to run otherwise untrusted code.
+The privileges and capabilities used in your TaskRun containers will also be present in your hermetic builds.
+Container escapes (and then network access) will always be possible.
+
+Hermetic builds should only be used to help detect and prevent accidental network access, and as an extra layer of defense against insider attacks. 
+These need not stop a determined adversary.
+
+### Risks and Mitigations
+
+One potential risk is that this namespace-level sandboxing is insufficient to make guarantees around hermeticity for most potential users of this feature.
+We can mitigate this by surveying users ahead of time, and we can always add more "strict" execution modes down the road.
+Note: We have verified that this will meet Google's requirements for hermetic builds of otherwise trusted code.
+
+
+### User Experience (optional)
+
+See the API changes outlined above.
+
+
+### Performance (optional)
+
+No performance implications are expected.
+
+## Design Details
+
+### Implementation
+
+See Alternatives Considered for a discussion of alternative approaches.
+
+The Tekton Pipelines entrypoint binary will be extended to support executing hermetic builds.
+This will consist of executing the user-controlled container entrypoint in a new Linux network namespace that is not configured.
+
+This requires passing some flags to `cmd.SysProcAttr.CloneFlags` before calling cmd.Run. 
+We will pass:
+
+```
+syscall.CLONE_NEWNET | syscall.CLONE_NEWUSER | syscall.CLONE_NEWNS | syscall.CLONE_NEW_PID
+```
+
+`CLONE_NEWNET` is the main flag we care about.
+This gives us a new, empty network namespace that will not work.
+To use this flag inside a Kubernetes container, we must also create a new user namespace using the other flags (unless the container runs with `privileged: True`). 
+
+After creating the user namespace, we must also map in the external users and groups.
+That can be done with the `cmd.SysProcAttr.UidMappings` and `cmd.SysProcAttr.GidMappings` fields, respectively. 
+
+The net result is a process running in a namespace with an identical filesystem, the same users and groups, and no networking.
+
+The entrypointer will need to know when to drop networking - we don't want to do this on every container!
+Specifically, init and resource container steps should still run with networking before and after the Task steps.
+We also only want to disable networking if we are running in a hermetic `ExecutionMode`.
+Unfortunately, the entrypointer does not know if it is running a Task step or a system container today.
+
+We'll add support for that via environment variables.
+
+### Implementation Plan
+
+In rough order:
+* Add experimental support for hermetic execution as an annotation everywhere possible (currently this is everywhere but PipelineRun.Spec.TaskRunSpecs)
```

Comment:
This is now possible today: https://godoc.org/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#EmbeddedTask

Reply:
Ooh perfect. Dropped that part.